src/nexmark/queries/q15.rs

```diff
@@ -90,122 +176,339 @@ pub fn q15(input: NexmarkStream) -> Q15Stream {
+        .distinct_incremental()
+        .index();
+
+    // Compute unique auctions across all bids and for each price range.
+    let distinct_auction = bids
+        .map(|(day, (auction, _price, _bidder))| (*day, *auction))
+        .distinct_incremental()
+        .index();
+    let rank1_distinct_auction = rank1_bids
+        .map(|(day, (auction, _price, _bidder))| (*day, *auction))
+        .distinct_incremental()
+        .index();
+    let rank2_distinct_auction = rank2_bids
+        .map(|(day, (auction, _price, _bidder))| (*day, *auction))
+        .distinct_incremental()
+        .index();
+    let rank3_distinct_auction = rank3_bids
+        .map(|(day, (auction, _price, _bidder))| (*day, *auction))
+        .distinct_incremental()
+        .index();
+
+    // Compute bids per day.
+    let count_total_bids: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> = bids
+        .index()
+        .aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank1_bids: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> = rank1_bids
+        .index()
+        .aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank2_bids: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> = rank2_bids
+        .index()
+        .aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank3_bids: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> = rank3_bids
+        .index()
+        .aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+
+    // Count unique bidders per day.
+    let count_total_bidders: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        distinct_bidder.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank1_bidders: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        rank1_distinct_bidder.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank2_bidders: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        rank2_distinct_bidder.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank3_bidders: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        rank3_distinct_bidder.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+
+    // Count unique auctions per day.
+    let count_total_auctions: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        distinct_auction.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank1_auctions: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        rank1_distinct_auction.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank2_auctions: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        rank2_distinct_auction.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+    let count_rank3_auctions: Stream<_, OrdIndexedZSet<OrdinalDate, isize, _>> =
+        rank3_distinct_auction.aggregate_linear::<(), _, _>(|_, _| -> isize { 1 });
+
+    // The following abomination simply joins all aggregates computed above into a
+    // single output stream.
```

Comment:
Yeah - wow. I would not have gotten to this point. I was initially looking for something like a `zip` function (or `zip_default`) before I ended up calculating the aggregates in the one loop.

Reply:
one day, we'll have mutiway joins.